Enable output padding when only outermost dim is dynamic

nandesuka opened the pull request:
Summary: When the shape of the output tensor has a dynamic outer most dim, the stride can still be padded to conform to configured alignment if required.

Test Plan:
CI

Rollback Plan:

Differential Revision: D79146886




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @mlazos

eellison commented on torch/_inductor/ir.py:
```diff
@@ -3707,9 +3707,10 @@ def _pad_strides(
         if not all(
             isinstance(s, (int, sympy.Integer))
-            for s in itertools.chain(in_strides, size)
+            for s in itertools.chain(in_strides, size[1:])
```
The stride that is not dynamic even when the the corresponding dimension is dynamic will be the least dense dimension, not necessarily the last dimension. 

It's only the last dimension when tensor is contiguous. 

Would it be sufficient to just check strides here 🤔

blaine-rister replied:
What about checking if `V.graph.sizevars.statically_known_equals(s, sympy.S.One)`? A good test case would be add -> transpose fusion. The output could still be padded even though its strides are permuted.

eellison replied:
Do we even need to check the shapes here ? If we check the strides, wouldn't that be encompass the above check ?

nandesuka replied:
Yup I think you are right we just need to check strides. I can take a look at spda in follow up. We probably should generalize the padding to all dimensions having dynamic shapes if possible.

eellison reviewed (commented):
Nice generalization but one comment.

Also, there's nothing preventing us from padding dynamic strides. If you look at the lowerings.py sdpa we pad to the inputs to dynamic aligned strides. 

The only thing to think abt is the heuristic in the lowering to not padding stride threshold.

blaine-rister commented on test/inductor/test_padding.py:
```diff
@@ -745,20 +746,11 @@ def get_input(size: tuple[int], alignment_bytes: int) -> torch.Tensor:
         input_tensors = [get_input(shape, alignment_bytes) for _ in range(num_inputs)]
 
         config_patches = {
-            "compile_threads": 1,
```
Is this just cleaning up unnecessary configs? Or is it related to the other changes?

nandesuka replied:
it's more of a clean up

blaine-rister commented:
Nice PR! It's cool that we can generalize padding to handle dynamic shapes.

blaine-rister commented on test/inductor/test_padding.py:
```diff
@@ -767,7 +759,54 @@ def get_input(size: tuple[int], alignment_bytes: int) -> torch.Tensor:
+        with config.patch(config_patches):
+            torch._dynamo.mark_dynamic(input_tensors[0], 0)
+            torch._dynamo.mark_dynamic(input_tensors[1], 0)
+            compiled = torch.compile(torch.add)
```
We might want to test a function with permute and/or transpose, just to make sure this generalizes to other stride orders. One example would be a 3D tensor where dims `(1,2)` are transposed, and dim 0 is dynamic.

blaine-rister reviewed (commented):
The core logic LGTM, but I left some comments about enhancing the test cases. I'll be ready to approve after that.

blaine-rister commented on test/inductor/test_padding.py:
```diff
@@ -767,7 +770,112 @@ def get_input(size: tuple[int], alignment_bytes: int) -> torch.Tensor:
+            ((32, 50, 30), 64, True),
+        ],
+    )
+    def test_outter_dynamic_shape_padding(self, shape, alignment_bytes, pad_output):
```
nit: typo
```suggestion
    def test_outer_dynamic_shape_padding(self, shape, alignment_bytes, pad_output):
```

blaine-rister commented on test/inductor/test_padding.py:
```diff
@@ -767,7 +770,112 @@ def get_input(size: tuple[int], alignment_bytes: int) -> torch.Tensor:
+        def get_input(size: tuple[int], alignment_bytes: int) -> torch.Tensor:
+            size_padded = list(size)
+            elem_size = 4  # float32
+            pad_elems = alignment_bytes // elem_size
```
nit: do we actually need input padding here? It seems like we're only testing that the output is padded.

blaine-rister commented on test/inductor/test_padding.py:
```diff
@@ -767,7 +770,112 @@ def get_input(size: tuple[int], alignment_bytes: int) -> torch.Tensor:
+        def get_input(size: tuple[int], alignment_bytes: int) -> torch.Tensor:
+            size_padded = list(size)
+            elem_size = 4  # float32
+            pad_elems = alignment_bytes // elem_size
```
Same comment as below. Do we need input padding for this test? There might be an opportunity to simplify the test case or reuse `get_padded_stride`.

blaine-rister reviewed (approved):
LGTM. I left a few tiny nits.

eellison commented on torch/_inductor/ir.py:
```diff
@@ -3725,11 +3723,11 @@ def _pad_strides(
-                stride = ceildiv(stride, align) * align
-                padded = True
-            new_strides[idx] = stride
+            if isinstance(stride, (int, sympy.Integer)):
```
this should definitively be an int right, since we checked for sympy.Integer above.

nandesuka replied:
We allow both int or sympy.int above. I am not sure if it's possible that the stride values are all sympy.int or partially int/sympy.int, prob doesn't hurt to leave it though.

eellison replied:
It should be an invariant based on our checks above. so it makes the code less clear

eellison reviewed (approved):
looks good one comment

nandesuka committed:
Enable output padding when only outermost dim is dynamic (#159404)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/159404

When the shape of the output tensor has a dynamic outer most dim, the stride can still be padded to conform to configured alignment if specified by padding config.

Test Plan:
CI

Rollback Plan:

Reviewed By: blaine-rister, eellison

Differential Revision: D79146886

nandesuka commented:
@pytorchbot label "release notes: inductor"

nandesuka commented:
@pytorchbot merge

nandesuka commented:
Abandoning in favour of: #160997
